Add GLIGEN Text Image implementation

tuanh123789 opened the pull request:
# What does this PR do?
  In #4441, there is a mention of the GLIGEN model, but I noticed that it doesn't include all the models referenced in the paper. In this Pull Request, I propose adding two additional models: Generation and Inpainting, with inputs including Text, Box, and particularly Image. With this model, users can input any object without the need to use Textual Inversion, Dreambooth, or LoRA.
All two checkpoint is converted from offical weight.


<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Core library:

- Schedulers: @williamberman and @patrickvonplaten
- Pipelines:  @patrickvonplaten and @sayakpaul
- Training examples: @sayakpaul and @patrickvonplaten
- Docs: @stevenliu and @yiyixu
- JAX and MPS: @pcuenca
- Audio: @sanchit-gandhi
- General functionalities: @patrickvonplaten and @sayakpaul

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- transformers: [different repo](https://github.com/huggingface/transformers)
- safetensors: [different repo](https://github.com/huggingface/safetensors)

-->

tuanh123789 commented:
Model checkpoint available on huggingface
Inpainting: anhnct/Gligen_Inpainting_Text_Image
Generation: anhnct/Gligen_Text_Image

sayakpaul commented:
Cc: @nikhil-masterful

tuanh123789 commented:
Can you check @nikhil-masterful

nikhil-masterful commented:
I'm a little under the weather right now, but will take a look over the weekend

patrickvonplaten commented:
Hey @tuanh123789 ,

Thanks a lot for your PR! Can we maybe put this pipeline in the community folder: https://github.com/huggingface/diffusers/tree/main/examples/community? 

We're sadly a bit overwhelmed by maintenance and aren't able to keep our slow tests clean - if we put this pipeline in the community folder, you can be the official author and the burden to maintain the pipeline is loosened. Would that be ok for you?

tuanh123789 commented:
> 

Thank you for responding. I understand everyone is busy, and it's okay to place this pipeline in the community folder. However, there's an issue: I'm not the author of this idea. I merely adapted the code and model weights from the original repository into a pipeline within Diffusers. Since this idea is based on a paper, similar to the approach in #4441, I believe it would be more accurate if we put it in https://github.com/huggingface/diffusers/tree/main/src/diffusers/pipelines/stable_diffusion. What are your thoughts on this?

sayakpaul commented:
Which use cases of GLIGEN are we missing?

tuanh123789 commented:
@sayakpaul 
In the GLIGEN paper, three main method types are mentioned:

Input box location, prompt, and phrases for the objects we want to appear inside the box.
Input box location, prompt, phrases, and an image of the specific object we want to appear inside the box.
Input canny map, semantic map, depth map, and prompt (quite similar to Controlnet).
In issue #4441, @nikhil-masterfull implemented the method mentioned in type 1 above. In this PR, I implemented the method from type 2. The advantage of this method is that you can place a real-world object or any image style into an image without needing methods like Textual Inversion, Dreambooth, or LoRA, you just need only reference image of each object or style. Additionally, you can introduce multiple objects into the image simultaneously and control where they appear by providing box location. It's important to note that methods 1 and 2 are separate pipelines, and PR #4441 is not missing.

![image](https://github.com/huggingface/diffusers/assets/52446297/86156d3c-ee5c-4bfc-bb96-eb8b2f83baf5)

> Which use cases of GLIGEN are we missing?

sayakpaul commented:
> The advantage of this method is that you can place a real-world object or any image style into an image without needing methods like Textual Inversion, Dreambooth, or LoRA, you just need only reference image of each object or style. Additionally, you can introduce multiple objects into the image simultaneously and control where they appear by providing box location.

This is very interesting. Could you provide a couple of samples for us to see and realize this advantage here? @patrickvonplaten considering this feature of some sort of zero-shot subject driven generation, I'd be in favor of graduating this pipeline in the core.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_4777). All of your documentation changes will be reflected on that endpoint.

tuanh123789 commented:
> > The advantage of this method is that you can place a real-world object or any image style into an image without needing methods like Textual Inversion, Dreambooth, or LoRA, you just need only reference image of each object or style. Additionally, you can introduce multiple objects into the image simultaneously and control where they appear by providing box location.
> 
> This is very interesting. Could you provide a couple of samples for us to see and realize this advantage here? @patrickvonplaten considering this feature of some sort of zero-shot subject driven generation, I'd be in favor of graduating this pipeline in the core.

Of course, here is some sample  with reference image: https://drive.google.com/drive/folders/1hIO-wvEoiTFWqCckdOgO7k0ilUniz9pQ?usp=sharing
Regarding the reference style, it seems like I have missed implementing some components. I think it's not too difficult, and I'm quite confident that I will complete it tomorrow.

tuanh123789 commented:
In my experience, using reference images with high resolution will help improve the quality of the generated images.

patrickvonplaten commented:
So far downloads have not be going up that much IMO:
https://huggingface.co/masterful/gligen-1-4-generation-text-box
(but trend is going upwards)

If you feel strongly about it @sayakpaul, let's add it to core (not 100% convinced at the moment, but happy to be proven wrong!)

tuanh123789 commented:
@sayakpaul 
I have updated the samples of the pipeline in the directory https://drive.google.com/drive/folders/1hIO-wvEoiTFWqCckdOgO7k0ilUniz9pQ?usp=sharing. Additionally, I have also implemented the missing features for style transfer. You can take a look, and I have also addressed the remaining errors. Could you please review it further? Thank you.

sayakpaul commented:
Wow! The results indeed seem quite amazing here. Thanks for your contributions!

sayakpaul commented on src/diffusers/models/attention.py:
```diff
@@ -154,7 +154,7 @@ def __init__(
 
         # 4. Fuser
-        if attention_type == "gated":
+        if attention_type == "gated" or attention_type == "GatedTextImage":
```
"gated_text_image" is a better name here IMO.

patrickvonplaten replied:
```suggestion
        if attention_type == "gated" or attention_type == "gated-text-image":
```
Can we stick to lowercase here? :-)

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_gligen_text_image.py:
```diff
@@ -0,0 +1,968 @@
+
+        >>> # Insert objects described by image at the region defined by bounding boxes
+        >>> pipe = StableDiffusionGLIGENTextImagePipeline.from_pretrained(
+        ...     "anhnct/Gligen_Inpainting_Text_Image", torch_dtype=torch.float16
```
These checkpoints might need regeneration if we modify the attention type value as suggested above. 

Also, let's ensure the checkpoints uploaded have thorough model cards.

sayakpaul replied:
Let's add some visual examples to the model cards as mentioned earlier :-)

sayakpaul commented:
Also, we won't be able to merge until and unless the CI is green.

tuanh123789 commented:
> Looks quite amazing to me.
> 
> * How did you convert the checkpoints to make them compatible with `diffusers`? If you had to perform any conversion, then the conversion script should also be in this PR.
> * Let's add a doc entry to https://huggingface.co/docs/diffusers/main/en/api/pipelines/stable_diffusion/gligen so that our users are aware of these capabilities.

tuanh123789 commented:
@sayakpaul 
To switch from the official checkpoint to diffusers, I've written a piece of code for conversion. I will add it to this PR along with adding a doc entry and a model card. I will inform you as soon as I complete it. Thank you very much.

tuanh123789 commented:
> 

I'll fix this too

nikhil-masterful commented:
@tuanh123789 @sayakpaul 

Why do we need a brand new pipeline `StableDiffusionGLIGENTextImagePipeline`  ?
Can we not add the functionality to the existing `StableDiffusionGLIGENPipeline` ?

This keeps the code clean and maintainable IMO

patrickvonplaten commented on src/diffusers/models/embeddings.py:
```diff
@@ -600,3 +600,61 @@ def forward(self, boxes, masks, positive_embeddings):
         return objs
+
+
+class PositionNetTextImage(nn.Module):
```
We need to add a docstring here describing where this class comes from

patrickvonplaten replied:
Actually instead of adding a new class here, can we put all the logic directly into `PositionNet` to remove duplicated logic? 

Could you add a new config variable `feature_type="text-only"  # "text-image"` here: https://github.com/huggingface/diffusers/blob/12358b986f15d0280f701034e42e9486253f2197/src/diffusers/models/embeddings.py#L566

It's important that we try to keep the `src/diffusers/models/embeddings` file as short as possible

patrickvonplaten commented on src/diffusers/models/unet_2d_condition.py:
```diff
@@ -571,8 +572,17 @@ def __init__(
+
             self.position_net = PositionNet(positive_len=positive_len, out_dim=cross_attention_dim)
 
+        elif attention_type == "GatedTextImage":
+            positive_len = 768
+            if isinstance(cross_attention_dim, int):
+                positive_len = cross_attention_dim
+            elif isinstance(cross_attention_dim, tuple) or isinstance(cross_attention_dim, list):
+                positive_len = cross_attention_dim[0]
+            self.position_net = PositionNetTextImage(positive_len=positive_len, out_dim=cross_attention_dim)
```
```suggestion
```

patrickvonplaten commented on src/diffusers/models/unet_2d_condition.py:
```diff
@@ -29,6 +29,7 @@
     ImageProjection,
     ImageTimeEmbedding,
     PositionNet,
+    PositionNetTextImage,
```
```suggestion
```

patrickvonplaten commented on src/diffusers/models/unet_2d_condition.py:
```diff
@@ -571,8 +572,17 @@ def __init__(
+
             self.position_net = PositionNet(positive_len=positive_len, out_dim=cross_attention_dim)
 
+        elif attention_type == "GatedTextImage":
```
```suggestion
```

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_4777). All of your documentation changes will be reflected on that endpoint.

patrickvonplaten commented on src/diffusers/models/unet_2d_condition.py:
```diff
@@ -571,8 +572,17 @@ def __init__(
             elif isinstance(cross_attention_dim, tuple) or isinstance(cross_attention_dim, list):
                 positive_len = cross_attention_dim[0]
+
             self.position_net = PositionNet(positive_len=positive_len, out_dim=cross_attention_dim)
```
Above replace:
```py
if attention_type == "gated":
```
with
```py
if attention_type in ["gated", "gated-text-image"]:
```
```suggestion
            self.position_net = PositionNet(positive_len=positive_len, out_dim=cross_attention_dim, feature_type="text" if attention_type == "gated" else "text-image")
```

patrickvonplaten commented on src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:
```diff
@@ -425,10 +425,8 @@ def run_safety_checker(self, image, device, dtype):
 
     def decode_latents(self, latents):
         warnings.warn(
-            (
```
Can we revert this? I think this messes up the quality check

patrickvonplaten commented on src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion_img2img.py:
```diff
@@ -426,10 +426,8 @@ def run_safety_checker(self, image, device, dtype):
 
     def decode_latents(self, latents):
         warnings.warn(
-            (
```
Can we revert this? I think this messes up the quality check

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion_img2img.py:
```diff
@@ -35,10 +35,8 @@
 # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion_img2img.preprocess with 8->64
 def preprocess(image):
     warnings.warn(
-        (
```
Can we revert this? I think this messes up the quality check

patrickvonplaten reviewed (commented):
Let's try to reduce complexity a bit in `unet_2d_condition.py` and `embeddings.py`

tuanh123789 commented:
> Let's try to reduce complexity a bit in `unet_2d_condition.py` and `embeddings.py`

@patrickvonplaten
Ok, I'll fix this. Thank you.

tuanh123789 commented:
Hi @patrickvonplaten, @sayakpaul 
I have updated the code section to adapt the official GLIGEN checkpoint for compatibility with Diffusers. Additionally, I have added a model card and further updated the documentation. Furthermore, I have also made code adjustments based on the reviews. Could anyone please check and review this for me? Thank you.

inpainting model card: https://huggingface.co/anhnct/Gligen_Inpainting_Text_Image
generate model card: https://huggingface.co/anhnct/Gligen_Text_Image

tuanh123789 commented:
> Thanks for your changes and contributions. If you could add some resultant images in the model card that would be amazing! Reviewing the PR now.

One more question, how do i fix check_repository_consistency error

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_gligen_text_image.py:
```diff
@@ -0,0 +1,979 @@
+        # We'll offload the last model manually.
+        self.final_offload_hook = hook
+
+    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline._encode_prompt
```
Could we incorporate the changes introduced in https://github.com/huggingface/diffusers/pull/4617?

https://github.com/huggingface/diffusers/blob/af3854d6ad5764163a99dfcf927f1e3541d5d738/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L253

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_gligen_text_image.py:
```diff
@@ -0,0 +1,979 @@
+            inpaint_mask[int(y0) : int(y1), int(x0) : int(x1)] = 0
+        return inpaint_mask
+
+    def crop(self, im, new_width, new_height):
+        width, height = im.size
+        left = (width - new_width) / 2
+        top = (height - new_height) / 2
+        right = (width + new_width) / 2
+        bottom = (height + new_height) / 2
+        return im.crop((left, top, right, bottom))
+
+    def target_size_center_crop(self, im, new_hw):
+        width, height = im.size
+        if width != height:
+            im = self.crop(im, min(height, width), min(height, width))
+        return im.resize((new_hw, new_hw), PIL.Image.LANCZOS)
```
Question for @yiyixuxu. Should these go to the image processor module? Or should the pipelines implement these on their own?

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_gligen_text_image.py:
```diff
@@ -0,0 +1,979 @@
+            im = self.crop(im, min(height, width), min(height, width))
+        return im.resize((new_hw, new_hw), PIL.Image.LANCZOS)
+
+    def project(self, x, hidden_size, device):
+        """
+        x (Batch*768) should be the penultimate feature of CLIP (before projection) projection_matrix (768*768) is the
+        CLIP projection matrix, which should be weight.data of Linear layer defined in CLIP (out_dim, in_dim), thus we
+        need to apply transpose below. this function will return the CLIP feature (without normalziation)
+        """
+        projection_matrix = hf_hub_download(repo_id="anhnct/Gligen_Text_Image", filename="projection_matrix")
+        projection_matrix = torch.load(projection_matrix).type(self.image_encoder.dtype).to(device).T
+        if hidden_size != projection_matrix.shape[0]:
+            return x
+        else:
+            return x @ torch.transpose(projection_matrix, 0, 1)
```
Not too happy about doing it this way. Can't we leverage the `hidden_states` of the CLIP text model and index the required features from there? @patrickvonplaten WDYT?

tuanh123789 replied:
In the official GLIGEN repository, they utilize the projection_matrix. In my implementation, I attempted to solely use the hidden_state of the CLIP model without employing the projection_matrix. However, upon comparing the outcomes, not utilizing the projection_matrix led to discrepancies in the results when compared to the official repository. If I use the projection_matrix here, the results produced in the original repository and the Diffusers implementation are the same.(Here, I used the same seed during inference)

sayakpaul replied:
Is this projection matrix not a part of the CLIP text model? If so, I think we should be able to leverage the matrix, no?

tuanh123789 replied:
They don't mention where the projection_matrix come from. They just provide a torch tensor of it.

tuanh123789 replied:
![image](https://github.com/huggingface/diffusers/assets/52446297/e1b6a410-29fd-4631-8249-e5066aeabb9c)
I think projection_matrix is not a part of CLIP model. They have it after training, i think so.

sayakpaul replied:
If that is the case, I think it makes sense to implement that as a separate `nn.Module` and keep that under the pipeline directory. We should then consider making that a part of the pipeline init i.e., a component of the pipeline (just like the UNet, text encoder, etc.). Maybe `projection` component. 

WDYT?

tuanh123789 replied:
I think that's quite reasonable, and I will proceed with implementing this idea.

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_gligen_text_image.py:
```diff
@@ -0,0 +1,979 @@
+            outputs = self.image_encoder(**inputs)
+            feature = outputs.image_embeds
+            feature = self.project(feature, hidden_size, device).squeeze(0)
+            feature = (feature / feature.norm()) * 28.7
```
Where does the magic number come from? I think it should be registered as a configuration variable.

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_gligen_text_image.py:
```diff
@@ -0,0 +1,979 @@
+            feature = outputs.pooler_output
+        return feature
+
+    def prepare_cross_attention_kwags(
```
```suggestion
    def prepare_cross_attention_kwargs(
```

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_gligen_text_image.py:
```diff
@@ -0,0 +1,979 @@
+
+        return out
+
+    def prepare_cross_attention_kwags_none(self, hidden_size, repeat_batch, max_objs, device):
```
Would prefer `kwargs`. How does this differ from the method above? I think having some brief docstrings would be very helpful here.

sayakpaul reviewed (commented):
Thanks for iterating. Left some comments. I am concerned about how `project()` is implemented. Apart from that:

* Let's ensure the involved methods (such as `enable_fuser`, `draw_inpaint_mask_from_boxes`, `complete_mask`, etc.) have brief docstrings. 
* Incorporate the latest changes for `encode_prompt()`. 
* Model cards have visual examples of the results. 

@stevhliu could you also scan the PR for the docs?

sayakpaul commented:
> One more question, how do i fix check_repository_consistency error

You can run `make fix-copies`.

tuanh123789 commented:
> Thanks for iterating. Left some comments. I am concerned about how `project()` is implemented. Apart from that:
> 
> * Let's ensure the involved methods (such as `enable_fuser`, `draw_inpaint_mask_from_boxes`, `complete_mask`, etc.) have brief docstrings.
> * Incorporate the latest changes for `encode_prompt()`.
> * Model cards have visual examples of the results.
> 
> @stevhliu could you also scan the PR for the docs?

OK, I will process those requests.

stevhliu commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_gligen_text_image.py:
```diff
@@ -0,0 +1,979 @@
+            cross_attention_kwargs (`dict`, *optional*):
+                A kwargs dictionary that if specified is passed along to the [`AttentionProcessor`] as defined in
+                [`self.processor`](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/attention_processor.py).
+            guidance_rescale (`float`, *optional*, defaults to 0.7):
```
Is `guidance_rescale` supported for this pipeline? I don't see it in the `__call__` signature.

tuanh123789 commented:
Hi @sayakpaul,
I have converted the projection_matrix to be a part of the pipeline and made it can load from checkpoint using `from_pretrained` method. Additionally, I have made the last changes to the `encode_prompt()` function, added brief docstrings, and included visual examples of results in the model card. Could you please review these? Thank you.

patrickvonplaten commented on src/diffusers/models/embeddings.py:
```diff
@@ -563,7 +565,7 @@ def __call__(self, x):
 
 class PositionNet(nn.Module):
-    def __init__(self, positive_len, out_dim, fourier_freqs=8):
+    def __init__(self, positive_len, out_dim, feature_type, fourier_freqs=8):
```
```suggestion
    def __init__(self, positive_len, out_dim, feature_type = "text-only", fourier_freqs=8):
```

for backwards compatibility

patrickvonplaten commented on src/diffusers/models/embeddings.py:
```diff
@@ -573,30 +575,81 @@ def __init__(self, positive_len, out_dim, fourier_freqs=8):
         return objs
+
+
+class CLIPImageProjection(ModelMixin, ConfigMixin):
```
We should not have model mixin classes in the `src/diffusers/models/embeddings.py` file

tuanh123789 replied:
Ok, I'll fix this

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/__init__.py:
```diff
@@ -42,10 +42,12 @@ class StableDiffusionPipelineOutput(BaseOutput):
 except OptionalDependencyNotAvailable:
     from ...utils.dummy_torch_and_transformers_objects import *  # noqa F403
 else:
+    from .clip_image_project_model import CLIPImageProjection
```
ok!

tuanh123789 commented:
> Looks good to me now! Ok to merge once @sayakpaul gives the 🟢 light!
> 
> Great job @tuanh123789 !

Thank you for support.

sayakpaul commented on src/diffusers/models/embeddings.py:
```diff
@@ -573,30 +573,70 @@ def __init__(self, positive_len, out_dim, fourier_freqs=8):
+                nn.Linear(512, out_dim),
+            )
+            self.linears_image = nn.Sequential(
+                nn.Linear(self.positive_len + self.position_dim, 512),
+                nn.SiLU(),
+                nn.Linear(512, 512),
+                nn.SiLU(),
+                nn.Linear(512, out_dim),
+            )
+            self.null_text_feature = torch.nn.Parameter(torch.zeros([self.positive_len]))
+            self.null_image_feature = torch.nn.Parameter(torch.zeros([self.positive_len]))
+
         self.null_position_feature = torch.nn.Parameter(torch.zeros([self.position_dim]))
 
-    def forward(self, boxes, masks, positive_embeddings):
+    def forward(
+        self,
+        boxes,
+        masks,
+        positive_embeddings=None,
+        phrases_masks=None,
+        image_masks=None,
+        phrases_embeddings=None,
+        image_embeddings=None,
+    ):
         masks = masks.unsqueeze(-1)
 
-        # embedding position (it may includes padding as placeholder)
-        xyxy_embedding = self.fourier_embedder(boxes)  # B*N*4 -> B*N*C
-
-        # learnable null embedding
-        positive_null = self.null_positive_feature.view(1, 1, -1)
+        xyxy_embedding = self.fourier_embedder(boxes)
```
Why are the comments going away?

sayakpaul commented on src/diffusers/models/embeddings.py:
```diff
@@ -573,30 +573,70 @@ def __init__(self, positive_len, out_dim, fourier_freqs=8):
+    def forward(
+        self,
+        boxes,
+        masks,
+        positive_embeddings=None,
+        phrases_masks=None,
+        image_masks=None,
+        phrases_embeddings=None,
+        image_embeddings=None,
+    ):
         masks = masks.unsqueeze(-1)
 
-        # embedding position (it may includes padding as placeholder)
-        xyxy_embedding = self.fourier_embedder(boxes)  # B*N*4 -> B*N*C
-
-        # learnable null embedding
-        positive_null = self.null_positive_feature.view(1, 1, -1)
+        xyxy_embedding = self.fourier_embedder(boxes)
         xyxy_null = self.null_position_feature.view(1, 1, -1)
-
-        # replace padding with learnable null embedding
-        positive_embeddings = positive_embeddings * masks + (1 - masks) * positive_null
         xyxy_embedding = xyxy_embedding * masks + (1 - masks) * xyxy_null
 
-        objs = self.linears(torch.cat([positive_embeddings, xyxy_embedding], dim=-1))
+        if positive_embeddings is not None:
+            positive_null = self.null_positive_feature.view(1, 1, -1)
+            positive_embeddings = positive_embeddings * masks + (1 - masks) * positive_null
+
+            objs = self.linears(torch.cat([positive_embeddings, xyxy_embedding], dim=-1))
```
Cool! I would still prefer to have the comments as this part of the code is a bit involved to navigate through.

sayakpaul commented on src/diffusers/models/unet_2d_condition.py:
```diff
@@ -565,13 +565,17 @@ def __init__(
         )
 
-        if attention_type == "gated":
+        if attention_type in ["gated", "gated-text-image"]:
```
Cool!

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_gligen_text_image.py:
```diff
@@ -0,0 +1,1034 @@
+            )
+        return image, has_nsfw_concept
+
+    def prepare_extra_step_kwargs(self, generator, eta):
```
Is it not copied from?

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_gligen_text_image.py:
```diff
@@ -0,0 +1,1034 @@
+            extra_step_kwargs["generator"] = generator
+        return extra_step_kwargs
+
+    def check_inputs(
```
No copy? If not, ignore the comment.

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_gligen_text_image.py:
```diff
@@ -0,0 +1,1034 @@
+
+        return out
+
+    def get_cross_attention_kwargs_without_grounded(self, hidden_size, repeat_batch, max_objs, device):
```
Much better name this one! Thank you.

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_gligen_text_image.py:
```diff
@@ -0,0 +1,1034 @@
+            inputs = self.processor(images=[input], return_tensors="pt").to(device)
+            outputs = self.image_encoder(**inputs)
+            feature = outputs.image_embeds
+            feature = self.image_project(feature).squeeze(0)
```
Sorry for iterating here a bit. 

We do have a projection class for the CLIP vision tower in `transformers`: https://huggingface.co/docs/transformers/model_doc/clip#transformers.CLIPVisionModelWithProjection. 

Is it possible to leverage pre-trained projection matrix and populate the projection layer of that class and serialize it? That way we won't need a separate class for that.

sayakpaul replied:
@patrickvonplaten WDYT?

sayakpaul reviewed (approved):
Thanks a lot for working on this one. Just left a final round of comments.

tuanh123789 commented:
Hi @sayakpaul 
Can you review the changes and can we merge. Thank you.

sayakpaul commented:
Thanks for your amazing contributions ❤️

patrickvonplaten commented:
Great work @tuanh123789 !
